### PR TITLE
Kotlin - Function names (including generics and extensions)

### LINKED ISSUE #792: Kotlin: function names are not highlighted
Example:
http://rouge.jneen.net/pastes/w8Wg

Note how function name is gray. Usually rouge highlights function names with another color. Compare with this Java example:

http://rouge.jneen.net/pastes/Mak3

I believe this is a regression of this:
#782

### PR DESCRIPTION
**Pull request by lordcodes**
Recognize Kotlin function names (declarations).

This includes handling extension functions, generic functions and generic extension functions.

The solution uses state to avoid long regexes.

- Function names mapped to Name.Function.
- Class name when it is an extension function mapped to Name.Class.
- Each generic parameter mapped to Name.Class.
- Less than (<), greater than (>) and comma (,) mapped to Punctuation.

Includes unit tests for each type.

Fixes #792, #791 and #906.

### CODE REVIEW

**Comment by lordcodes** [comment]
CC @dblessing

**Comment by lordcodes** [comment]
@jneen Sorry to bump this PR, but would be really keen to get these Kotlin PRs I have put forward included, as currently Rouge is almost unusable for Kotlin as it doesn't really highlight anything other than certain keywords.

File: spec/lexers/kotlin_spec.rb [hunk 1]
```diff
@@ -22,5 +22,77 @@
...
+          ['Name.Class', 'ClassT'],
+          ['Punctuation', '>'],
+          ['Text', " "],
+          ['Name.Class', 'String'],
+          ['Punctuation', '.'],
+          ['Name.Function', 'makeField'],
+          ['Punctuation', '()']
+      end
+
+    end
```
Comment (dblessing) [inline spec/lexers/kotlin_spec.rb]: Same here as in #1001. Please remove these, but thank you for the sentiment. If there aren't already examples covering this in samples, please add those as well.

Comment (lordcodes) [reply spec/lexers/kotlin_spec.rb]: Tests removed and adding missing samples.

**Comment by lordcodes** [comment]
@dblessing Removed the tests and added missing samples. Thanks for review.

**Pull request merged**